sdks/python/apache_beam/runners/direct/evaluation_context.py

```diff
@@ -69,67 +69,109 @@ class _SideInputsContainer(object):
-    with self._lock:
-      view = self._views[side_input]
-      assert not view.has_result
-      assert view.value is None
-      assert view.callable_queue is not None
-      view.value = self._pvalue_to_value(side_input, view.elements)
-      view.elements = None
-      result = tuple(view.callable_queue)
-      for task in result:
-        task.blocked = False
-      view.callable_queue = None
-      view.has_result = True
-      return result
-
-  def update_watermarks_for_transform(self, ptransform, watermark):
-    # Collect tasks that get unblocked as the workflow progresses.
-    unblocked_tasks = []
-    for view in self._transform_to_views[ptransform]:
-      unblocked_tasks.extend(self._update_watermarks_for_view(view, watermark))
-    return unblocked_tasks
+  def update_watermarks_for_transform_and_unblock_tasks(self,
+                                                        ptransform,
+                                                        watermark):
+    """Updates _SideInputsContainer after a watermark update and unbloks tasks.
+
+    It traverses the list of side inputs per PTransform and calls
+    _update_watermarks_for_side_input_and_unblock_tasks to unblock tasks.
+
+    Args:
+      ptransform: Value of a PTransform.
+      watermark: Value of the watermark after an update for a PTransform.
 
-  def _update_watermarks_for_view(self, view, watermark):
+    Returns:
+      Tasks that get unblocked as a result of the watermark advancing.
+    """
     unblocked_tasks = []
-    if watermark.input_watermark == WatermarkManager.WATERMARK_POS_INF:
-      unblocked_tasks = self.finalize_value_and_get_tasks(view)
+    for side in self._transform_to_side_inputs[ptransform]:
+      unblocked_tasks.extend(
+          self._update_watermarks_for_side_input_and_unblock_tasks(
+              side, watermark))
     return unblocked_tasks
 
-  def _pvalue_to_value(self, view, values):
-    """Given a side input view, returns the associated value in requested form.
+  def _update_watermarks_for_side_input_and_unblock_tasks(self,
+                                                          side_input,
+                                                          watermark):
+    """Helps update _SideInputsContainer after a watermark update.
+
+    For each view of the side input, it updates the value of the watermark
+    recorded when the watermark moved and unblocks tasks accordingly.
+
+    Args:
+      side_input: (_UnpickledSideInput) value.
+      watermark: Value of the watermark after an update for a PTransform.
+
+    Returns:
+      Tasks that get unblocked as a result of the watermark advancing.
+    """
+    with self._lock:
+      unblocked_tasks = []
+      view = self._views[side_input]
+      view.watermark = watermark
+
+      view = self._views[side_input]
+      tasks_just_unblocked = []
+      for task, block_until in view.blocked_tasks:
+        if watermark.input_watermark >= block_until:
```

Comment:
Question for me to understand, since this is the watermark associated with the view, would not we care about blocking until the output watermark passes block until point?

Reply:
If your question is why the '=' in '>=' is necessary, the reason is that we need to unblock tasks when the watermark reaches infinity (and not after). Hence my choice of the name (when you reach `block_until`, block no more). 
It would have been possible to define the unblocking point as `unblock_after` and in that case we would anyway need to make the infinity case work, by doing:
```
if watermark.input_watermark > unblock_after or watermark.input_watermark == WatermarkManager.WATERMARK_POS_INF:
```

Reply:
I think the question is regarding the input vs output watermark, and which should be used here.  I think Ahmet is right here in that we should use the output watermark for this comparison.  The input watermark describes the lateness of inputs to this step, and the output watermark describes the lateness of outputs.  My understanding is that the previous side input code was not correct in this respect, since it incorrectly compared against the input watermark.

Reply:
There are two usages of input_watermark in the code added by this PR: at L104 and L161. I have printed both input_watermark and output_watermark at those points with the existing tests and their values are the same. Can you give me an example where they could be different, so that I can check if we are not testing something needed?